Java sprint 15

1EVILGUN1 opened the pull request:
Добавление класса Booking , Доработка dto выделение под конкретные задачи, Подключение бд postgresql, добавление запросов в repository JPA

1EVILGUN1 committed:
Java sprint 15

Добавление класса Booking , Доработка dto выделение под конкретные задачи, Подключение бд postgresql, добавление запросов в repository JPA

VadimZharkov commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -0,0 +1,67 @@
+    private final BookingService service;
+
+    @PostMapping
+    public Booking add(@RequestHeader("X-Sharer-User-Id") Long userId, @Validated @RequestBody BookingAddDto dto) {
```
Здесь нужно `@Valid` использовать.
Посмотри в статье в чем разница
https://www.baeldung.com/spring-valid-vs-validated

VadimZharkov commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -0,0 +1,67 @@
+
+    private final BookingService service;
+
+    @PostMapping
```
Возвращать из метода контроллера нужно BookingDto

VadimZharkov commented on src/main/java/ru/practicum/shareit/booking/service/BookingService.java:
```diff
@@ -0,0 +1,30 @@
+
+import java.util.List;
+
+@Transactional(readOnly = true)
```
Лучше разместить на классе реализации, т.к. данные аннотации относятся к деталям реализации
https://stackoverflow.com/questions/3120143/where-should-i-put-transactional-annotation-at-an-interface-definition-or-at-a

VadimZharkov commented on src/main/java/ru/practicum/shareit/error/ErrorHandler.java:
```diff
@@ -0,0 +1,66 @@
+
+    @ExceptionHandler({MethodArgumentNotValidException.class, ValidationException.class, NotAvailableException.class})
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public Map<String, String> handleValidationException(final Exception e) {
```
Лучше возвращать специально созданный для сообщений об ошибках тип ErrorResponse, а не "обезличенный" map

VadimZharkov commented on src/main/java/ru/practicum/shareit/item/controller/ItemController.java:
Те же замечания, что и для booking. 
Отдаем в методах контроллера Dto и `@Valid `вместо `@Validated`

VadimZharkov commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -1,12 +1,45 @@
-    private Long userId;
+
+    @NotBlank(message = "Название не может быть пустым")
+    @NotNull(message = "Название не может быть Null")
```
`@NotBlank`  уже содержит проверку на null

VadimZharkov reviewed (changes requested):
В целом хорошая работа, посмотри что нужно исправить и можно улучшить

1EVILGUN1 committed:
fix error

Исправлены замечания ревью, добавлены улучшение читаемости в  BookingService и изменение класса ErrorResponse на record

1EVILGUN1 committed:
update errorresponse

аннотация getter
